src/sentry/rules/processing/buffer_processing.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import math
+import uuid
+from abc import ABC, abstractmethod
+from dataclasses import asdict, dataclass
+from datetime import datetime, timezone
+from itertools import islice
+from typing import ClassVar
+
+from celery import Task
+
+from sentry import buffer, options
+from sentry.buffer.redis import BufferHookEvent, redis_buffer_registry
+from sentry.db import models
+from sentry.utils import metrics
+from sentry.utils.registry import NoRegistrationExistsError, Registry
+
+logger = logging.getLogger("sentry.delayed_processing")
+
+
+@dataclass
+class FilterKeys:
+    project_id: int
+
+
+@dataclass
+class BufferHashKeys:
+    model: type[models.Model]
+    filters: FilterKeys
+
+
+class DelayedProcessingBase(ABC):
+    buffer_key: ClassVar[str]
+
+    def __init__(self, project_id: int):
+        self.project_id = project_id
+
+    @property
+    @abstractmethod
+    def hash_args(self) -> BufferHashKeys:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def processing_task(self) -> Task:
+        raise NotImplementedError
+
+
+delayed_processing_registry = Registry[type[DelayedProcessingBase]]()
+
+
+def fetch_alertgroup_to_event_data(
+    project_id: int, model: type[models.Model], batch_key: str | None = None
+) -> dict[str, str]:
+    field: dict[str, models.Model | int | str] = {
+        "project_id": project_id,
+    }
+
+    if batch_key:
+        field["batch_key"] = batch_key
+
+    return buffer.backend.get_hash(model=model, field=field)
+
+
+def bucket_num_groups(num_groups: int) -> str:
+    if num_groups > 1:
+        magnitude = 10 ** int(math.log10(num_groups))
+        return f">{magnitude}"
+    return "1"
+
+
+def process_in_batches(project_id: int, processing_type: str) -> None:
+    """
+    This will check the number of alertgroup_to_event_data items in the Redis buffer for a project.
+
+    If the number is larger than the batch size, it will chunk the items and process them in batches.
+
+    The batches are replicated into a new redis hash with a unique filter (a uuid) to identify the batch.
+    We need to use a UUID because these batches can be created in multiple processes and we need to ensure
+    uniqueness across all of them for the centralized redis buffer. The batches are stored in redis because
+    we shouldn't pass objects that need to be pickled and 10k items could be problematic in the celery tasks
+    as arguments could be problematic. Finally, we can't use a pagination system on the data because
+    redis doesn't maintain the sort order of the hash keys.
+
+    `processing_task` will fetch the batch from redis and process the rules.
+    """
+    batch_size = options.get("delayed_processing.batch_size")
```

Comment:
Should this file also provide a `DEFAULT_BATCH_SIZE` constant that we use as the default value here? (I'm guessing we also set that in the options tho 😅)

Reply:
hmm this can be a constant, yes. but i'll keep it as reading the option

Reply:
i think it should be local var because i'm not sure how often the entire file is reloaded? updating options doesn't trigger pods to restart

Reply:
sorry, i was just thinking: `batch_size = options.get("delayed_processing.batch_size", DEFAULT_BATCH_SIZE)`

That way we could have a module definition and see what we'd expect it to be about(ish) or that we'd need to look at the option. not a huge issue to add though so nbd to skip.

Reply:
there is a default in the option declaration
https://github.com/getsentry/sentry/blob/e28e4b66808018449b142aca4ec8d1571dca09eb/src/sentry/options/defaults.py#L2818-L2822